### DISCUSSION TITLE
I can`t run RPMSG-lite in Zephyr 3.3 in imx7d-pico/m4 side.  Zephyr 1.13 Works. Issue in new IPM changes in Zephyr

### DISCUSSION THREAD

**Discussion by neuberfran** [zephyrproject-rtos/zephyr#66669] [General]
![porting](https://github.com/zephyrproject-rtos/zephyr/assets/38701397/523b5a21-5b85-4393-8662-7b1831fe1a0d)

Contextualizing:

1) I resolved the issues(Pic above) I had immediately when trying to migrate (west init, west update, west export) a project from zephyr 1.13 to zephyr 3.3.

![porting2](https://github.com/zephyrproject-rtos/zephyr/assets/38701397/6ec30f22-5bd2-4eed-a086-ed59346684c9)
2) Since then (as shown in the photo above) zephyr.bin (from zephyr3.3) has been defined to start on m4 of imx7d-pico in  U-boot in TCM memory 0X7F80000 and the messages  from main_remote_echo (in M4 UART6-Console is: Waiting for Master) stop at while (!rpmsg_lite_is_link_up(rl_instance)) {
**github project:** https://github.com/neuberfran/diegosueiro/blob/main/rpmsg-lite/zephyr/samples/subsys/ipc/rpmsg_lite/remote_echo/src/main_remote_echo.c#L82

Some things I've already tried to resolve:
1) Changes in Yocto/A7/Hardknott
2) Changes in Zephyr Side, in DTS files, Kconfig files, Soc.c file, Exclusions of things about libmetal and open-amp that could block RPMSG-lite, etc...
3) Changes zephyr_base And now I have rl_instance # 0, But issue now in this two lines **528 and 549:**
https://github.com/neuberfran/diegosueiro/blob/main/rpmsg-lite/lib/rpmsg_lite/rpmsg_lite.c#L549

in line 529 rpmsg_lite_dev and non-zero

But, in line 548 rpmsg_lite_dev->link_state is always equal to zero (and return 0 to function (!rpmsg_lite_is_link_up(rl_instance)). Why?

How to resolve my issue(Pls)?

![porting3](https://github.com/zephyrproject-rtos/zephyr/assets/38701397/583cd6db-82b1-441d-96e8-817aae1a9d00)

**Comment by MichalPrincNXP** [comment]
Hi @neuberfran 
if I understand it correctly, you are attempting to run the older rpmsg-lite v2.0.0 based example with the latest Zephyr version, correct? What is the boot order, is the Linux starting before the Zephyr or vice versa? When the Zephyr app stucks in rpmsg_lite_is_link_up(), I would check if all MU (Messaging Unit) interrupts are reached on the Zephyr side, have you tried to place a breakpoint into the MU isr?
We do not have the imx7d-pico board to try but we could take another NXP board with CortexA and CortexM cores and could get some rpmsg_lite example running. This will take some time, expecting some outputs in first weeks of 2024, does it make sense to you and is it acceptable?
Regards
Michal

**Comment by neuberfran** [comment]
@MichalPrincNXP Yes, I stay trying older rpmsg-lite based example with the latest Zephyr version. I don't know if older rpmsg-lite version is 2.0.0.

The boot order is the traditional one in these cases. U-boot starts zephyr.bin (some messages appear on the m4 side and stop at that damn moment)

Also, I looked at the issue below 1 month ago and tried not to upload the tty module directly to  scorpion.conf but after booting Linux/A7/Yocto Hardknott using modprobe imx_rpmsg_tty. But not solved.
https://github.com/EmbeddedRPC/erpc-imx-demos/issues/2#issuecomment-375855176

You said:
I would check if all MU (Messaging Unit) interrupts are reached on the Zephyr side, have you tried to place a breakpoint into the MU isr?
Could you give more details on how to do this?

Today I will try this here:
https://docs.zephyrproject.org/latest/samples/drivers/ipm/ipm_imx/README.html#ipm-imx
https://review.coreboot.org/plugins/gitiles/zephyr-cros/+/7ca69b385de7896fc83ffe86723290c9a4417416%5E%21/

How unfortunate that you don't have this device. Would anything device/kit with A7 and M4 produce the same results(to You test)?

Yes, I can wait a little longer because I still want to know what is stopping rpmsg-lite here.

Could you tell me something about lines 528 and 549 that I asked above?

E.g.: I stay trying rpmsg-lite (old and new version) in FreeRTOS here. With the new RPMSG-lite version I will can use something like this: rpmsg_lite_wait_for_link_up(my_rpmsg, RL_BLOCK);
But I have a compilation error

With Old rpmsg-lite version in FreeRTOS, stay stopping in rpmsg_lite_is_link_up() like here.

root@scorpion:~# **ls /sys/bus/virtio/devices/**
virtio0
root@scorpion:~# **ls /sys/bus/rpmsg/devices**
virtio0.rpmsg_ctrl.0.0	virtio0.rpmsg_ns.53.53
root@scorpion:~# **dmesg | grep -E "MU| rpmsg"**
[    0.635986] imx rpmsg driver is registered.
[    0.637319] MU is ready for cross core communication!
[    0.699771] virtio_rpmsg_bus virtio0: rpmsg host is online
[    0.716669] hw perfevents: enabled with armv7_cortex_a7 PMU driver, 5 counters available
root@scorpion:~# uname -a
Linux scorpion 5.10.72-2.2.0-next+g000eb85263cb #1 SMP PREEMPT Thu Jan 12 09:09:37 UTC 2023 armv7l armv7l armv7l GNU/Linux

**Reply by neuberfran** [reply]
@MichalPrincNXP 
I had said I would try ipm_imx app:
https://review.coreboot.org/plugins/gitiles/zephyr-cros/+/7ca69b385de7896fc83ffe86723290c9a4417416%5E%21/

https://docs.zephyrproject.org/latest/samples/drivers/ipm/ipm_imx/README.html#ipm-imx

**I have issue here in  west build -b pico_pi_m4  samples/drivers/ipm/ipm_imx/**  And
**west build -p auto -b pico_pi_m4 remote_echo** Using new ipm_imx.c And device.h Drivers.

https://gist.github.com/neuberfran/7fea2adcc911556b3c5ae70c16668f30

I have some indications that my issue may be in TSC new way of dealing with ipm_imx (ipm.h driver):
https://github.com/zephyrproject-rtos/zephyr/pull/38523#issuecomment-919376434
https://github.com/zephyrproject-rtos/zephyr/issues/38609

**in old_project (zephyr1.13)**
 find . -name ipm.h
./rpmsg-lite/zephyr/samples/subsys/ipc/rpmsg_lite/remote_echo/build/pico_pi_m4/zephyr/include/generated/syscalls/ipm.h
./zephyr/manifest-tmp/include/zephyr/drivers/ipm.h
./zephyr/include/ipm.h

![Captura de tela de 2023-12-19 19-10-19](https://github.com/zephyrproject-rtos/zephyr/assets/38701397/69ab16f0-97e8-40eb-9360-94b8c3db2678)

**Reply by neuberfran** [reply]
@MichalPrincNXP You can Test RPMS-lite in Zephyr Upstream using evkmcimx7ul (https://github.com/neuberfran/pmsg_lite_files/tree/main/boards/evkmcimx7ulp) by MCXpresso/NXP and/or in imx7s-Warp7 (NXP\)
https://www.youtube.com/watch?v=4zMyjaEgj2U&t=1210s

**Comment by MichalPrincNXP** [comment]
Hello @neuberfran 
as for the MU isr checking, I meant to place breakpoint into platform_ipm_callback() that should be called when the MU interrupt is raised.

Anyway, I would propose to wait a little ... we are going to support rpmsg_lite in Zephyr more at the beginning on the next year. We internally already get Zephyr rpmsg_lite examples running on different NXP dualcore platforms using the latest Zephyr code and the latest Zephyr IPC infrastructure. We focused on dualcores with CortexM cores so far, checking Zephyr rpmsg-lite communication against Linux is planned as well (your case). Based on [this records](https://docs.zephyrproject.org/latest/boards/index.html) the proposed evkmcimx7ulp board is not in the scope of officially supported Zephyr boards, so we will probably choose another NXP board to test. However, porting to your board of choice should be very easy once we finish. Does it make sense and helpful for you?

**Comment by neuberfran** [comment]
@MichalPrincNXP Ok. But Why I can't work with old rpmsg_lite in FreeRTOS 1.0.1?

https://forums.freertos.org/t/issue-when-transpose-rpmsg-lite-project-from-zephyr-1-13-to-freertos-1-0-1/18698/2

In FreeRTOS 1.0.1 I can Run normal RPMSG

**Comment by MichalPrincNXP** [comment]
Hi @neuberfran , have you made any progress in your rpmsg_lite based project? 
As for running rpmsg_lite in FreeRTOS:
- is it really needed to use the old FreeRTOS 1.0.1 version?
- it is not needed to use the rpmsg_platform_zephyr_ipm.c alternative in FreeRTOS env., when running the rpmsg_lite in FreeRTOS you have to include lib/rpmsg_lite/porting/environment/rpmsg_env_freertos.c, lib/include/environment/freertos/rpmsg_env_specific.h, lib/include/rpmsg_queue.h, lib/rpmsg_lite/rpmsg_queue.c sources in your project. 
- I would recommend to download the NXP MCUXpressoSDK package for the evkmcimx7ulp from [https://mcuxpresso.nxp.com](https://mcuxpresso.nxp.com) and look at the boards\evkmcimx7ulp\multicore_examples\rpmsg_lite_str_echo_rtos\ example that is FreeRTOS based

**Reply by neuberfran** [reply]
Off course, in FreeRTOS project I will use rpmsg_env_freertos.c, etc...  And Not rpmsg_platform_zephyr_ipm.c

**Comment by neuberfran** [comment]
@MichalPrincNXP 

Before answering your questions, let me say that I am eagerly waiting to have the RPMSG-lite in zephyr upstream on my imx7d-pico (technexion) device.

About what I said above, today I'm trying to upload zephyr.elf (coming from the upstream RPMSG-lite zephyr) in userspace using remoteproc and maybe I'll leave this place (rpmsg_lite_is_link_up(rl_instance)) { ) where the upstream RPMSG-lite has stopped here
https://www.youtube.com/watch?v=NFyEJXIpq4E
https://community.nxp.com/t5/i-MX-Processors/i-MX8MP-and-FreeRTOS-stuck-at-rpmsg-lite-is-link-up-still/td-p/1415356

Regarding your questions, I say that it is not essential to use FreeRTOS 1.0.1 but any suitable FreeRTOS

I'm a FAN (and I've already reproduced it here: https://www.youtube.com/watch?v=tNevT71nCpM) of this Timada project: 
https://qiita.com/t-imada/items/f34ff62d3124a3950423?s=03
Where he used a new FreeRTOS and said on github that I can use Remoteproc on my device to control (start, stop) my m4-side in userspace: https://github.com/TImada/raspi4_freertos_rpmsg/issues/1


I don't have this device (evkmcimx7ulp) But I compile mcuxpresso here Ok.  
The problems start when I try to transpose this project to imx7d-pico. The imx7d-pico uses UART6 on the M4 side (and also on the Yocto hardknott) and the evkmcimx7ulp uses UART0. I even (in my own way) managed to play the project for UART6 and it compiled. But the messages do not appear on the screen on the m4 side when running in U-boot

**Reply by neuberfran** [reply]
@MichalPrincNXP Hello. How is the rpmsg-lite project on the upstream zephyr?

**Comment by MichalPrincNXP** [comment]
the rpmsg-lite gitrepo update should be available on Jan-24

**Reply by neuberfran** [reply]
@MichalPrincNXP Hi, Is it possible to run rpmsg_lite (Linux for Zephyr) on Zephyr Upstream?

**Reply by neuberfran** [reply]
@MichalPrincNXP 
![Captura de tela de 2023-12-22 13-48-18](https://github.com/zephyrproject-rtos/zephyr/assets/38701397/6448e19d-7d45-426f-a0a7-e6a3259378de)

https://docs.zephyrproject.org/latest/samples/drivers/ipm/ipm_imx/README.html#ipm-imx

west build -b udoo_neo_full_m4 samples/drivers/ipm/ipm_imx
This command above only works with this board (It would be a good hello_world for our project, do you agree?)
Why does it only work with udoo_neo_full_m4?

**Reply by MichalPrincNXP** [reply]
As for the Zephyr support, rpmsg_lite is not yet part of the upstream, but it has been provided within the rpmsg_lite Github repo, see https://github.com/nxp-mcuxpresso/rpmsg-lite/releases/tag/v5.1.1. 
Please look at the Zephyr folder and read the description here: https://github.com/nxp-mcuxpresso/rpmsg-lite/tree/main/zephyr
more info can be provided by @TomasGalbickaNXP 
As for udoo neo, I am not familiar with that and unfortunatelly I can't help you on that.

**Reply by TomasGalbickaNXP** [reply]
Hi, @neuberfran definitely good start to try. 
I do not have these boards but at least build works on my side. Could you try latest Zephyr main?

Please see my log for versions: [v3.6.0-rc1-39-g3cbe0b33f3b6]
```
west build -b colibri_imx7d_m4 samples/drivers/ipm/ipm_imx --build-dir build-ipm_imx -p
-- west build: making build dir /home/tom/zephyrproject/zephyr/build-ipm_imx pristine
-- west build: generating a build system
Loading Zephyr default modules (Zephyr base).
-- Application: /home/tom/zephyrproject/zephyr/samples/drivers/ipm/ipm_imx
-- CMake version: 3.22.1
-- Found Python3: /home/tom/zephyrproject/.venv/bin/python3 (found suitable version "3.10.12", minimum required is "3.8") found components: Interpreter
-- Cache files will be written to: /home/tom/.cache/zephyr
-- Zephyr version: 3.6.0-rc1 (/home/tom/zephyrproject/zephyr)
-- Found west (found suitable version "1.2.0", minimum required is "0.14.0")
-- Board: colibri_imx7d_m4
-- ZEPHYR_TOOLCHAIN_VARIANT not set, trying to locate Zephyr SDK
-- Found host-tools: zephyr 0.16.4 (/home/tom/zephyr-sdk-0.16.4)
-- Found toolchain: zephyr 0.16.4 (/home/tom/zephyr-sdk-0.16.4)
-- Found Dtc: /home/tom/zephyr-sdk-0.16.4/sysroots/x86_64-pokysdk-linux/usr/bin/dtc (found suitable version "1.6.0", minimum required is "1.4.6")
-- Found BOARD.dts: /home/tom/zephyrproject/zephyr/boards/arm/colibri_imx7d_m4/colibri_imx7d_m4.dts
-- Generated zephyr.dts: /home/tom/zephyrproject/zephyr/build-ipm_imx/zephyr/zephyr.dts
-- Generated devicetree_generated.h: /home/tom/zephyrproject/zephyr/build-ipm_imx/zephyr/include/generated/devicetree_generated.h
-- Including generated dts.cmake file: /home/tom/zephyrproject/zephyr/build-ipm_imx/zephyr/dts.cmake
Parsing /home/tom/zephyrproject/zephyr/Kconfig
Loaded configuration '/home/tom/zephyrproject/zephyr/boards/arm/colibri_imx7d_m4/colibri_imx7d_m4_defconfig'
Merged configuration '/home/tom/zephyrproject/zephyr/samples/drivers/ipm/ipm_imx/prj.conf'
Configuration saved to '/home/tom/zephyrproject/zephyr/build-ipm_imx/zephyr/.config'
Kconfig header saved to '/home/tom/zephyrproject/zephyr/build-ipm_imx/zephyr/include/generated/autoconf.h'
-- Found GnuLd: /home/tom/zephyr-sdk-0.16.4/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.2.0/../../../../arm-zephyr-eabi/bin/ld.bfd (found version "2.38")
-- The C compiler identification is GNU 12.2.0
-- The CXX compiler identification is GNU 12.2.0
-- The ASM compiler identification is GNU
-- Found assembler: /home/tom/zephyr-sdk-0.16.4/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc
-- Using ccache: /usr/bin/ccache
CMake Warning at /home/tom/zephyrproject/zephyr/CMakeLists.txt:864 (message):
  No SOURCES given to Zephyr library: drivers__clock_control

  Excluding target from build.


-- Configuring done
-- Generating done
-- Build files have been written to: /home/tom/zephyrproject/zephyr/build-ipm_imx
-- west build: building application
[1/118] Preparing syscall dependency handling

[2/118] Generating include/generated/version.h
-- Zephyr version: 3.6.0-rc1 (/home/tom/zephyrproject/zephyr), build: v3.6.0-rc1-39-g3cbe0b33f3b6
[117/118] Linking C executable zephyr/zephyr.elf
Memory region         Used Size  Region Size  %age Used
           FLASH:       12784 B        32 KB     39.01%
             RAM:        3808 B        32 KB     11.62%
        IDT_LIST:          0 GB        32 KB      0.00%
Generating files from /home/tom/zephyrproject/zephyr/build-ipm_imx/zephyr/zephyr.elf for board: colibri_imx7d_m4
[118/118] cd /home/tom/zephyrproject/zephyr/build-ipm_imx/zephyr && /home/tom....py --elf-file=/home/tom/zephyrproject/zephyr/build-ipm_imx/zephyr/zephyr.elf
```

**Reply by neuberfran** [reply]
@TomasGalbickaNXP I can Now with west build -b colibri_imx7d_m4  samples/drivers/ipm/ipm_imx --build-dir build-ipm_imx -p but I cant with west build -b pico_pi_m4 samples/drivers/ipm/ipm_imx --build-dir build-ipm_imx -p (I have the same errors as in the photo)

```
-- CMake version: 3.22.1
-- Found Python3: /usr/bin/python3 (found suitable version "3.10.12", minimum required is "3.8") found components: Interpreter 
-- Cache files will be written to: /home/neuberfran/.cache/zephyr
-- Zephyr version: 3.6.0-rc1 (/home/neuberfran/zephyrproject2/zephyr)
-- Found west (found suitable version "1.0.0", minimum required is "0.14.0")
-- Board: colibri_imx7d_m4
-- ZEPHYR_TOOLCHAIN_VARIANT not set, trying to locate Zephyr SDK
-- Found host-tools: zephyr 0.16.1 (/home/neuberfran/zephyr-sdk-0.16.1)
-- Found toolchain: zephyr 0.16.1 (/home/neuberfran/zephyr-sdk-0.16.1)
-- Found Dtc: /home/neuberfran/zephyr-sdk-0.16.1/sysroots/x86_64-pokysdk-linux/usr/bin/dtc (found suitable version "1.6.0", minimum required is "1.4.6") 
-- Found BOARD.dts: /home/neuberfran/zephyrproject2/zephyr/boards/arm/colibri_imx7d_m4/colibri_imx7d_m4.dts
```
Issue:
```
west build -b pico_pi_m4  samples/drivers/ipm/ipm_imx --build-dir build-ipm_imx -p
-- west build: generating a build system
Loading Zephyr default modules (Zephyr base).
-- Application: /home/neuberfran/zephyrproject2/zephyr/samples/drivers/ipm/ipm_imx
-- CMake version: 3.22.1
-- Found Python3: /usr/bin/python3 (found suitable version "3.10.12", minimum required is "3.8") found components: Interpreter 
-- Cache files will be written to: /home/neuberfran/.cache/zephyr
-- Zephyr version: 3.6.0-rc1 (/home/neuberfran/zephyrproject2/zephyr)
-- Found west (found suitable version "1.0.0", minimum required is "0.14.0")
-- Board: pico_pi_m4
-- ZEPHYR_TOOLCHAIN_VARIANT not set, trying to locate Zephyr SDK
-- Found host-tools: zephyr 0.16.1 (/home/neuberfran/zephyr-sdk-0.16.1)
-- Found toolchain: zephyr 0.16.1 (/home/neuberfran/zephyr-sdk-0.16.1)
-- Found Dtc: /home/neuberfran/zephyr-sdk-0.16.1/sysroots/x86_64-pokysdk-linux/usr/bin/dtc (found suitable version "1.6.0", minimum required is "1.4.6") 
-- Found BOARD.dts: /home/neuberfran/zephyrproject2/zephyr/boards/arm/pico_pi_m4/pico_pi_m4.dts
-- Generated zephyr.dts: /home/neuberfran/zephyrproject2/zephyr/build-ipm_imx/zephyr/zephyr.dts
-- Generated devicetree_generated.h: /home/neuberfran/zephyrproject2/zephyr/build-ipm_imx/zephyr/include/generated/devicetree_generated.h
-- Including generated dts.cmake file: /home/neuberfran/zephyrproject2/zephyr/build-ipm_imx/zephyr/dts.cmake

warning: The choice symbol IPM_IMX_MAX_DATA_SIZE_4 (defined at drivers/ipm/Kconfig.imx:26) was
selected (set =y), but no symbol ended up as the choice selection. See
http://docs.zephyrproject.org/latest/kconfig.html#CONFIG_IPM_IMX_MAX_DATA_SIZE_4 and/or look up
IPM_IMX_MAX_DATA_SIZE_4 in the menuconfig/guiconfig interface. The Application Development Primer,
Setting Configuration Values, and Kconfig - Tips and Best Practices sections of the manual might be
helpful too.

Parsing /home/neuberfran/zephyrproject2/zephyr/Kconfig
Loaded configuration '/home/neuberfran/zephyrproject2/zephyr/boards/arm/pico_pi_m4/pico_pi_m4_defconfig'
Merged configuration '/home/neuberfran/zephyrproject2/zephyr/samples/drivers/ipm/ipm_imx/prj.conf'
Configuration saved to '/home/neuberfran/zephyrproject2/zephyr/build-ipm_imx/zephyr/.config'
Kconfig header saved to '/home/neuberfran/zephyrproject2/zephyr/build-ipm_imx/zephyr/include/generated/autoconf.h'
-- Found GnuLd: /home/neuberfran/zephyr-sdk-0.16.1/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.2.0/../../../../arm-zephyr-eabi/bin/ld.bfd (found version "2.38") 
-- The C compiler identification is GNU 12.2.0
-- The CXX compiler identification is GNU 12.2.0
-- The ASM compiler identification is GNU
-- Found assembler: /home/neuberfran/zephyr-sdk-0.16.1/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc
CMake Warning at /home/neuberfran/zephyrproject2/zephyr/CMakeLists.txt:864 (message):
  No SOURCES given to Zephyr library: drivers__clock_control

  Excluding target from build.


CMake Warning at /home/neuberfran/zephyrproject2/zephyr/CMakeLists.txt:864 (message):
  No SOURCES given to Zephyr library: drivers__ipm

  Excluding target from build.


-- Configuring done
-- Generating done
-- Build files have been written to: /home/neuberfran/zephyrproject2/zephyr/build-ipm_imx
-- west build: building application
[1/112] Preparing syscall dependency handling

[2/112] Generating include/generated/version.h
-- Zephyr version: 3.6.0-rc1 (/home/neuberfran/zephyrproject2/zephyr), build: v3.6.0-rc1-70-g505cc19941bb
[21/112] Building C object CMakeFiles/app.dir/src/main.c.obj
FAILED: CMakeFiles/app.dir/src/main.c.obj 
/home/neuberfran/zephyr-sdk-0.16.1/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc -DKERNEL -DK_HEAP_MEM_POOL_SIZE=0 -DPICOLIBC_LONG_LONG_PRINTF_SCANF -D__LINUX_ERRNO_EXTENSIONS__ -D__PROGRAM_START -D__ZEPHYR__=1 -I/home/neuberfran/zephyrproject2/zephyr/drivers -I/home/neuberfran/zephyrproject2/zephyr/include -I/home/neuberfran/zephyrproject2/zephyr/build-ipm_imx/zephyr/include/generated -I/home/neuberfran/zephyrproject2/zephyr/soc/arm/nxp_imx/mcimx7_m4 -I/home/neuberfran/zephyrproject2/modules/hal/cmsis/CMSIS/Core/Include -I/home/neuberfran/zephyrproject2/zephyr/modules/cmsis/. -I/home/neuberfran/zephyrproject2/modules/hal/nxp/imx/drivers/. -I/home/neuberfran/zephyrproject2/modules/hal/nxp/imx/devices/. -I/home/neuberfran/zephyrproject2/modules/hal/nxp/imx/devices/MCIMX7D/. -I/home/neuberfran/zephyrproject2/zephyr/modules/hal_nxp/. -isystem /home/neuberfran/zephyrproject2/zephyr/lib/libc/common/include -fno-strict-aliasing -Os -imacros /home/neuberfran/zephyrproject2/zephyr/build-ipm_imx/zephyr/include/generated/autoconf.h -fno-printf-return-value -fno-common -g -gdwarf-4 -fdiagnostics-color=always -mcpu=cortex-m4 -mthumb -mabi=aapcs -mfp16-format=ieee -mtp=soft --sysroot=/home/neuberfran/zephyr-sdk-0.16.1/arm-zephyr-eabi/arm-zephyr-eabi -imacros /home/neuberfran/zephyrproject2/zephyr/include/zephyr/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wdouble-promotion -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-unused-but-set-variable -Werror=implicit-int -fno-pic -fno-pie -fno-asynchronous-unwind-tables -ftls-model=local-exec -fno-reorder-functions --param=min-pagesize=0 -fno-defer-pop -fmacro-prefix-map=/home/neuberfran/zephyrproject2/zephyr/samples/drivers/ipm/ipm_imx=CMAKE_SOURCE_DIR -fmacro-prefix-map=/home/neuberfran/zephyrproject2/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/home/neuberfran/zephyrproject2=WEST_TOPDIR -ffunction-sections -fdata-sections --specs=picolibc.specs -std=c99 -MD -MT CMakeFiles/app.dir/src/main.c.obj -MF CMakeFiles/app.dir/src/main.c.obj.d -o CMakeFiles/app.dir/src/main.c.obj -c /home/neuberfran/zephyrproject2/zephyr/samples/drivers/ipm/ipm_imx/src/main.c
/home/neuberfran/zephyrproject2/zephyr/samples/drivers/ipm/ipm_imx/src/main.c: In function 'ipm_callback':
/home/neuberfran/zephyrproject2/zephyr/samples/drivers/ipm/ipm_imx/src/main.c:20:26: error: 'CONFIG_IPM_IMX_MAX_DATA_SIZE' undeclared (first use in this function)
   20 |         for (i = 0; i < (CONFIG_IPM_IMX_MAX_DATA_SIZE / 4); i++) {
      |                          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/neuberfran/zephyrproject2/zephyr/samples/drivers/ipm/ipm_imx/src/main.c:20:26: note: each undeclared identifier is reported only once for each function it appears in
In file included from /home/neuberfran/zephyrproject2/zephyr/include/zephyr/toolchain/gcc.h:98,
                 from /home/neuberfran/zephyrproject2/zephyr/include/zephyr/toolchain.h:50,
                 from /home/neuberfran/zephyrproject2/zephyr/include/zephyr/kernel_includes.h:23,
                 from /home/neuberfran/zephyrproject2/zephyr/include/zephyr/kernel.h:17,
                 from /home/neuberfran/zephyrproject2/zephyr/samples/drivers/ipm/ipm_imx/src/main.c:7:
/home/neuberfran/zephyrproject2/zephyr/samples/drivers/ipm/ipm_imx/src/main.c: In function 'main':
/home/neuberfran/zephyrproject2/zephyr/include/zephyr/device.h:89:41: error: '__device_dts_ord_204' undeclared (first use in this function); did you mean '__device_dts_ord_24'?
   89 | #define DEVICE_NAME_GET(dev_id) _CONCAT(__device_, dev_id)
      |                                         ^~~~~~~~~
/home/neuberfran/zephyrproject2/zephyr/include/zephyr/toolchain/common.h:137:26: note: in definition of macro '_DO_CONCAT'
  137 | #define _DO_CONCAT(x, y) x ## y
      |                          ^
/home/neuberfran/zephyrproject2/zephyr/include/zephyr/device.h:89:33: note: in expansion of macro '_CONCAT'
   89 | #define DEVICE_NAME_GET(dev_id) _CONCAT(__device_, dev_id)
      |                                 ^~~~~~~
/home/neuberfran/zephyrproject2/zephyr/include/zephyr/device.h:216:37: note: in expansion of macro 'DEVICE_NAME_GET'
  216 | #define DEVICE_DT_NAME_GET(node_id) DEVICE_NAME_GET(Z_DEVICE_DT_DEV_ID(node_id))
      |                                     ^~~~~~~~~~~~~~~
/home/neuberfran/zephyrproject2/zephyr/include/zephyr/device.h:233:34: note: in expansion of macro 'DEVICE_DT_NAME_GET'
  233 | #define DEVICE_DT_GET(node_id) (&DEVICE_DT_NAME_GET(node_id))
      |                                  ^~~~~~~~~~~~~~~~~~
/home/neuberfran/zephyrproject2/zephyr/samples/drivers/ipm/ipm_imx/src/main.c:36:15: note: in expansion of macro 'DEVICE_DT_GET'
   36 |         ipm = DEVICE_DT_GET(DT_NODELABEL(mub));
      |               ^~~~~~~~~~~~~
/home/neuberfran/zephyrproject2/zephyr/samples/drivers/ipm/ipm_imx/src/main.c:44:17: error: 'CONFIG_IPM_IMX_MAX_DATA_SIZE' undeclared (first use in this function)
   44 |                 CONFIG_IPM_IMX_MAX_DATA_SIZE);
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
[30/112] Building C object zephyr/arch...eFiles/arch__arm__core.dir/fatal.c.obj
ninja: build stopped: subcommand failed.
FATAL ERROR: command exited with status 1: /usr/bin/cmake --build /home/neuberfran/zephyrproject2/zephyr/build-ipm_imx
neuberfran@ubuntu2204:~/zephyrproject2/zephyr$ 
```

If you put rpmsg-lite (new version) in this board: **colibri_imx7d_m4**  (Yocto in A7 Side and Zephyr Upstream in M4 Side) Possibly, I will also get it on my board: **pico_pi_m4**

**Reply by TomasGalbickaNXP** [reply]
Please add this board file to support `pico_pi_m4`. It compiles afterwards.

file: `samples/drivers/ipm/ipm_imx/boards/pico_pi_m4.overlay`

```
/*
 * Copyright 2024 NXP
 *
 * SPDX-License-Identifier: Apache-2.0
 */

&mub {
	status = "okay";
};
```

Patch:
```patch
diff --git a/samples/drivers/ipm/ipm_imx/boards/pico_pi_m4.overlay b/samples/drivers/ipm/ipm_imx/boards/pico_pi_m4.overlay
new file mode 100644
index 0000000000..2174d931a1
--- /dev/null
+++ b/samples/drivers/ipm/ipm_imx/boards/pico_pi_m4.overlay
@@ -0,0 +1,9 @@
+/*
+ * Copyright 2024 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+&mub {
+	status = "okay";
+};
```

**Reply by neuberfran** [reply]
@TomasGalbickaNXP 

Ok. Now I have zephyr.bin with mub
Note: I'm already preparing a Yocto Kirkstone image with the m4 part prepared with mub, but:
Since the link below shown in the article to prepare the Linux part is not working, I would like you to inform me if the generated zephyr.bin is to be placed in Uboot and what should I do in UserSpace to test the ipm_imx?
https://source.codeaurora.org/external/imxsupport/imx-mu/
![Captura de tela de 2024-02-08 15-00-22](https://github.com/zephyrproject-rtos/zephyr/assets/38701397/5d02e25a-bc71-46b3-9ae4-48300d278d51)
https://gist.github.com/neuberfran/e3ed5a0c010cd04f170571860fc11ba0

**Reply by neuberfran** [reply]
@TomasGalbickaNXP Kernel Panic ! in everytime I use mub (and Not with mu). WHy?

**3 device tree files here:**
(https://gist.github.com/neuberfran/87e4ca4fc40e2d95f28cf89d1dce857b)

I prepare Uboot (2022.04) with:
setenv loadm4image 'fatload mmc ${mmcdev}:${mmcpart} 80008000 ${m4image}; cp.b 80008000 ${tcm_addr} 8000'
**And**
setenv m4image zephyr.bin
saveenv
reset

![Screenshot from 2024-02-08 15-35-30](https://github.com/zephyrproject-rtos/zephyr/assets/38701397/4a482a55-3856-424a-ac65-d877f2367441)
![Screenshot from 2024-02-08 15-36-03](https://github.com/zephyrproject-rtos/zephyr/assets/38701397/66cc5b30-d6ed-4f9e-80e7-802ff3e4c267)

**Comment by TomasGalbickaNXP** [comment]
Hello @neuberfran currently the RPMSG-Lite under Zephyr with Linux as main core is in plan, once there will be example on this you can take is as reference for your board.

And as stated in https://github.com/zephyrproject-rtos/zephyr/discussions/66669#discussioncomment-8013100
`I would recommend to download the NXP MCUXpressoSDK package for the evkmcimx7ulp from https://mcuxpresso.nxp.com/ and look at the boards\evkmcimx7ulp\multicore_examples\rpmsg_lite_str_echo_rtos\ example that is FreeRTOS based`
It would be better to use that example to communicate with Linux via RPMSG-Lite.
Thanks.

**Comment by neuberfran** [comment]
https://github.com/zephyrproject-rtos/zephyr/issues/69433
